### PR TITLE
Resolve #1456 : Make QueryPlanResult and QueryPlanInfo immutable.

### LINKED ISSUE #1456: Make QueryPlanResult immutable
In order to prevent concurrency issues with the components of the QueryPlanResult, the class should be made immutable, where from a certain point, no changes could be made to it or to its components.

### PR DESCRIPTION
**Pull request by ohadzeliger**
make Make QueryPlanResult and QueryPlanInfo immutable:
- Create builder for QueryPlanInfo
- Remove setter methods from QueryPlanInfo
- Add constructor with PlanInfo to QueryPlanResult

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanInfo.java [hunk 1]
```diff
@@ -141,4 +139,43 @@ public String toString() {
...
+         * @return this
+         */
+        @Nonnull
+        public <T> Builder put(@Nonnull QueryPlanInfoKey<T> key, @Nonnull T value) {
+            infoMap.put(key, value);
+            return this;
+        }
+
+        @Nullable
+        public <T> T get(@Nonnull QueryPlanInfoKey<T> key) {
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanInfo.java:173]: Normally, builders don't have gets -- do you need this one?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanResult.java [hunk 2]
```diff
@@ -39,7 +39,12 @@
 
     public QueryPlanResult(@Nonnull final RecordQueryPlan plan) {
         this.plan = plan;
-        planInfo = new QueryPlanInfo();
+        planInfo = QueryPlanInfo.EMPTY;
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanResult.java:42]: Usually, it's better to have a static method `empty()` on `QueryPlanInfo` to return `EMPTY` when called. This allows you to encapsulate whether `QueryPlanInfo` really should just return the same instance over and over or whether you create a different instance every time. It's a pattern I follow but that's just nitpicking.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanInfo.java (hunk 1 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanInfo.java:173]: Yes, as it turns out, with the usage flow, we needed to log the contents of the PlanInfo before it is built, hence added a `get()`. It seemed to not interfere with immutability.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanResult.java (hunk 2 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/QueryPlanResult.java:42]: That makes sense. Done.

**Review by normen662 — APPROVED** [review]

**Pull request merged**